cpp/include/cudf/transform.hpp

```diff
@@ -0,0 +1,44 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef CUDF_TRANSFORM_HPP
+#define CUDF_TRANSFORM_HPP
+
+#include "cudf.h"
+#include "types.h"
+
+namespace cudf {
+
+/**---------------------------------------------------------------------------*
+ * @brief Creates a new column by applying a unary function against every
+ * element of an input column.
+ *
+ * Computes:
+ * `out[i] = F(in[i])`
+ *
+ * @param input               The input column to transform
+ * @param ptx_unary_function  The PTX of the unary function to apply
+ * @param outout_type         The output type that is compatible with the output type in the PTX code
+ * @return gdf_column         The column resulting from applying the unary function to
+ *                            every element of the input
+ *---------------------------------------------------------------------------**/
+gdf_column transform(const gdf_column &input,
```

Comment:
This requires `rdc`.

Reply:
Unless you're talking about providing template function.

Reply:
I think @harrism meant the functionality to directly accept the result of parsing the PTX code, i.e a CUDA function with explicit types.

Reply:
> I think @harrism meant the functionality to directly accept the result of parsing the PTX code, i.e a CUDA function with explicit types.

Unless you're talking about accepting a functor (meaning a template function). What you're suggesting would require accepting a `__device__` function pointer, which would mean enabling relocatable device code.

Reply:
I think I misunderstood the original suggestion, which I understand now to mean accept a _string_ containing a C++ device function. In which case, that is fine and does not require `rdc`.

Reply:
Actually, you understood me correctly. But I had forgotten about this requiring RDC. Still on my list  to have someone investigate the cost of RDC to libcudf perf...  Maybe once we have a benchmark suite.

However, I do like the idea of allowing C++ API users to use the JIT-based transform API by providing a string containing a C++ device function!

Reply:
Actually, I initially thought that we'd let people write UDFs in CUDA rather than using numba generated PTX. That makes more sense and [CuPy also has it](https://docs-cupy.chainer.org/en/latest/tutorial/kernel.html). Just a general comment, not asking for it rn.

Reply:
I have added the option to feed a CUDA string, in addition to PTX string option.